uw_oidc/id_token.py

```diff
@@ -1,56 +1,98 @@
-                          key=key,
-                          algorithms=self.SIGNING_ALGORITHMS,
-                          issuer=getattr(settings, 'TOKEN_ISSUER', 'uwidp'),
-                          audience=getattr(settings, 'TOKEN_AUDIENCE', ''),
-                          leeway=int(getattr(settings, 'TOKEN_LEEWAY', 1)))
-
-        except InvalidSignatureError as ex:
-            if self.get_key(force_update=True) != key:
-                return self.decode_token()
+            headers = get_unverified_header(self.token)
+        except PyJWTError as ex:
+            raise InvalidTokenHeader(ex)
 
-            raise InvalidTokenError(ex)
+        if headers.get('kid') is None or not len(headers['kid']):
+            logger.error("InvalidTokenHeader: missing kid {}".format(headers))
+            raise InvalidTokenHeader()
 
-        except PyJWTError as ex:
-            raise InvalidTokenError(ex)
+        return headers['kid']
 
-    def username_from_token(self):
-        self.key_id = self.extract_keyid()
-        return self.decode_token().get('sub')
+    def get_token_payload(self, refresh_keys=False):
+        """
+        Raise InvalidTokenError if not a valid jwt token.
+        """
+        pubkey = self.get_key(refresh_keys)
+        if pubkey is None:
+            if refresh_keys is False:
+                return self.get_token_payload(refresh_keys=True)
+            logger.error("NoMatchingPublicKey (kid: {})".format(self.key_id))
+            raise NoMatchingPublicKey(self.key_id)
 
-    def extract_keyid(self):
-        # TODO
-        return "defaultEC"
+        # When reaching this point, we have got the valid public key.
+        try:
+            return self.decode_token(pubkey)
+        except PyJWTError as ex:
+            logger.error("InvalidTokenError {} {}".format(ex, self.token))
+            raise InvalidTokenError(ex)
 
-    def get_key(self, force_update=False):
+    def get_key(self, force_update):
         return UWIdPToken.JWKS_CLIENT.get_pubkey(
             self.key_id, force_update=force_update)
+
+    def decode_token(self, pubkey):
+        return decode(
+            self.token,
+            options=self.JWT_OPTIONS,
+            key=pubkey,
+            algorithms=self.SIGNING_ALGORITHMS,
+            issuer=getattr(settings, 'UW_TOKEN_ISSUER',
+                           "https://idp-eval.u.washington.edu"),
+            audience=getattr(settings, 'UW_TOKEN_AUDIENCE'),
+            leeway=int(getattr(settings, 'UW_TOKEN_LEEWAY', 1)))
+
+    def valid_auth_time(self):
+        """
+        Raise InvalidTokenError if the id-token is not fresh enough
+        """
+        timeout = int(getattr(settings, 'UW_TOKEN_MAX_AGE', 0))
+        if timeout == 0:
+            # will not check the auth_time
+            return True
+        age_limit = timegm(datetime.utcnow().utctimetuple()) - timeout
+        try:
+            return int(self.payload['auth_time']) >= age_limit
```

Comment:
If we decide to set session expiration and refresh the tokens, I will make a separate PR to check the "iat" (token issue time) attribute instead of "auth_time".